TheEthicalHackerCup/Assets/Scripts/GameManager.cs

```diff
@@ -0,0 +1,130 @@
+    public static event Action<int> OnReputationChange;
+    public static event Action<int> OnOpponentKnowledgeChange;
+    public static event Action<SecurityConcepts, int> OnDefenseUpgradeLevelsChange;
+    public static event Action<SecurityConcepts, int> OnAttackSpecificHeatChange;
+    public static event Action<object> OnPassiveAttackLogChange; //TODO: Update with passive attack type Enum from HG
+    public static event Action<SecurityConcepts> OnIncommingAttackLogChange;
+    /*
+        HOW TO SUBSCRIBE TO EVENTS
+        
+        // Consider this class the observable, to make a different area an observer, do this in an initializer method
+        GameManager.OnReputationChange += LocallyDefinedFunction_ParamMatchingAction
+        // In this case, LocallyDefinedFunction_ParamMatchingAction(int reputation) would be a defined function
+
+        // Make sure to make an OnDestroy() method to unsubscribe from the event like this 
+        GameManager.OnReputationChange -= LocallyDefinedFunction_ParamMatchingAction
+    */
+
+
+    // Called before application starts as the script loads
+    void Awake()
+    {
+        // Destroy instance if it somehow already exists and is not this
+        if (Instance != null && Instance != this) { Destroy(this); }
+        else { Instance = this; }
+
+        Instance._reputation = 0;
+        Instance._opponentKnowledge = 0;
+        Instance._defenseUpgradeLevels = new Dictionary<SecurityConcepts, int>();
+        Instance._attackSpecificHeat = new Dictionary<SecurityConcepts, int>();
+        Instance._passiveAttackLog = new List<object>(); //TODO: Update with passive attack type Enum from HG
+        Instance._incommingAttackLog = new List<SecurityConcepts>();
+
+        // Iterate thru security concepts to instantiate zeros for defense upgrade and attack heat
+        foreach (SecurityConcepts concept in Enum.GetValues(typeof(SecurityConcepts)))
+        {
+            _defenseUpgradeLevels.Add(concept, 0);
+            _attackSpecificHeat.Add(concept, 0);
+        }
+
+    }
+
+    /*
+        HOW TO CALL THESE METHODS
+        GameManager.Instance.SetReputation(20);
+        MAKE SURE TO USE THE INSTANCE PROPERTY AS THAT IS THE SINGLE INSTANCE
+    */
+
+
+    // Getters
+    public int GetReputation() { return _reputation; }
+    public int GetOpponentKnowledge() { return _opponentKnowledge; }
+    public Dictionary<SecurityConcepts, int> GetDefenseUpgradeLevels() { return _defenseUpgradeLevels; }
+    public Dictionary<SecurityConcepts, int> GetAttackSpecificHeat() { return _attackSpecificHeat; }
+    public List<object> GetPassiveAttackLog()
+    {
+        //TODO: Update with passive attack type Enum from HG
+        return _passiveAttackLog;
+    }
+    public List<SecurityConcepts> GetIncommingAttackLog() { return _incommingAttackLog; }
+
+
+    // Primitive Setters
+    public void SetReputation(int reputation)
+    {
+        if (reputation > 0 && reputation < MaxReputation)
```

Comment:
what about repuatation  = 0. Don't we start at that value?

Reply:
I feel like we should have a incrementer method so that we hide the implementation details of set and get reputation from callers

Reply:
Updated to allow for zero

Since reputation and opp knowledge can both go up and down, set would be easier to use than incrementors and decrementors. 

Also note that there will be events emitted when these values get changed so anyone who needs these values wouldn't need to constantly call the getter, they would just need to subscribe themselves as an observer. 

Keeping the getters around would be useful for areas that don't need constant access to the game state. Perhaps that could be things like the end of game score screen which would just need to get the values once to display

Reply:
Updated to allow for zero

Since reputation and opp knowledge can both go up and down, set would be easier to use than incrementors and decrementors.

Also note that there will be events emitted when these values get changed so anyone who needs these values wouldn't need to constantly call the getter, they would just need to subscribe themselves as an observer.

Keeping the getters around would be useful for areas that don't need constant access to the game state. Perhaps that could be things like the end of game score screen which would just need to get the values once to display